File: src/main/go/leetcode/lists/hot100/LeetCode39.go
package hot100

func combinationSum(candidates []int, target int) [][]int {
	if len(candidates) == 0 {
		return [][]int{}
	}
	var res [][]int
	var dfs func([]int, []int, int, int, int, *[][]int)
	dfs = func(candidates []int, temp []int, index int, sum int, target int, res *[][]int) {
		if target < sum {
			return
		}
		if target == sum {
			*res = append(*res, append([]int(nil), temp...))
			return
		}
		for i := index; i < len(candidates); i++ {
			temp = append(temp, candidates[i])
			sum = sum + candidates[i]
			dfs(candidates, temp, i, sum, target, res)
			temp = temp[:len(temp)-1]
			sum = sum - candidates[i]
		}
	}
	dfs(candidates, []int{}, 0, 0, target, &res)
	return res
}

File: src/main/java/leetcode/lists/hot100/LeetCode39.java
package leetcode.lists.hot100;

import java.util.ArrayList;
import java.util.List;

/**
 * @author aiwujingxin@gmail.com
 * @date 2022/9/11 18:32
 */
public class LeetCode39 {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        if (candidates == null || candidates.length == 0) {
            return new ArrayList<>();
        }
        List<List<Integer>> res = new ArrayList<>();
        dfs(candidates, 0, 0, target, res, new ArrayList<>());
        return res;
    }

    private void dfs(int[] candidates, int index, int sum, int target, List<List<Integer>> res, List<Integer> list) {
        if (target < sum) {
            return;
        }

        if (target == sum) {
            res.add(new ArrayList<>(list));
            return;
        }

        for (int i = index; i < candidates.length; i++) {
            list.add(candidates[i]);
            sum += candidates[i];
            dfs(candidates, i, sum, target, res, list);
            sum -= candidates[i];
            list.remove(list.size() - 1);
        }
    }
}
